data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,711 @@
+
+### Currying and Uncurrying
+
+In the previous sections, two kinds of multiple-parameter functions have been presented.
+- Functions returning a function, such as `sweet_cat` and `sour_cat`
+- Functions taking a tuple as a parameter, such as `spicy_cat`
+
+Interestingly, both kinds of functions provide a way to pass several pieces of data to a body while being functions with a single parameter. From this perspective, it makes sense to say: “All functions have a single argument.”
+
+This goes even further. It is always possible to translate back and forth between functions that look like `sweet_cat` (or `sour_cat`) and functions that look like `spicy_cat`.
+
+These translations have names:
+* [Currying](https://en.wikipedia.org/wiki/Currying) goes from the `spicy_cat` form into the `sour_cat` (or `sweet_cat`) form.
+* Uncurrying goes from the `sour_cat` (or `sweet_cat`) form into the `spicy_cat` form.
+
+It also said that `sweet_cat` and `sour_cat` are _curried_ functions whilst `spicy_cat` is _uncurried_.
+
+These translations are attributed to the 20th-century logician [Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry).
+
+From a typing perspective, this means the following types are equivalent:
+- `string -> (string -> string)` &mdash; curried function type
+- `string * string -> string` &mdash; uncurried function type
+
+Here, this is shown using `string` as an example, but it applies to any group of three types.
+
+We will not dive any deeper into the details here, but these equivalences can be formally defined using _ad-hoc_ mathematics.
+
+Often, currying or uncurrying are applied when refactoring the code. One form is changed into the other as an edit. However, it is also possible to implement one from the other to have both forms available:
+```ocaml
+# let uncurried_cat (x, y) = sweet_cat x y;;
+val uncurried_cat : string * string -> string = <fun>
+
+# let curried_cat x y = uncurried_cat (x, y);;
+val curried_cat : string -> string -> string = <fun>
+```
+
+Functions can take several equivalent forms, depending on the way data is passed or returned. It is important to:
+* Be able to refactor in any direction
+* Have both forms starting from any of each.
+
+Currying and uncurrying should be understood as operations acting on functions the same way addition and substractions are operations acting on numbers.
+
+In practice, curried functions are the default form functions should take because:
+- They allow partial application
+- It is less editing, no parentheses or commas
+- No pattern matching over a tuple takes place
+
+## Functions With Side-Effects
+
+With respect to its type, a function is processes input data from its *domain*, the set of input values, and produce a result data from its *codomain*, the set of output values.
+```ocaml
+# string_of_int;;
+- : int -> string = <fun>
+```
+The function `string_of_int`
+* Has domain `int`
+* Has codomain `string`
```

Comment:
```suggestion
* Has domain `int` (left of `->` / input data)
* Has codomain `string` (right of `->` / output data)
```

These terms are to help avoid saying the "type at the right" or "type at the left" of a function's type arrow.